cmd/diff/main.go

```diff
@@ -0,0 +1,290 @@
+}
+
+type fileNodeDiff struct {
+	path string
+	from bufcas.Digest
+	to   bufcas.Digest
+	diff string
+}
+
+type manifestDiff struct {
+	addedPaths         map[string]bufcas.FileNode
+	removedPaths       map[string]bufcas.FileNode
+	changedDigestPaths map[string]fileNodeDiff
+}
+
+func newManifestDiff() *manifestDiff {
+	return &manifestDiff{
+		addedPaths:         make(map[string]bufcas.FileNode),
+		removedPaths:       make(map[string]bufcas.FileNode),
+		changedDigestPaths: make(map[string]fileNodeDiff),
+	}
+}
+
+func (d *manifestDiff) added(path string, node bufcas.FileNode) {
+	d.addedPaths[path] = node
+}
+
+func (d *manifestDiff) removed(path string, node bufcas.FileNode) {
+	d.removedPaths[path] = node
+}
+
+func (d *manifestDiff) changedDigest(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+	path string,
+	fromNode bufcas.FileNode,
+	toNode bufcas.FileNode,
+) error {
+	if fromNode.Path() != toNode.Path() {
+		return errors.New("from path and to path are different")
+	}
+	fromData, err := storage.ReadPath(ctx, bucket, fileNodePath(fromNode))
+	if err != nil {
+		return fmt.Errorf("read from changed path: %w", err)
+	}
+	toData, err := storage.ReadPath(ctx, bucket, fileNodePath(toNode))
+	if err != nil {
+		return fmt.Errorf("read to changed path: %w", err)
+	}
+	diffData, err := diff.Diff(
+		ctx,
+		fromData,
+		toData,
+		fromNode.Digest().String(),
+		toNode.Digest().String(),
+		diff.DiffWithSuppressCommands(),
+		diff.DiffWithSuppressTimestamps(),
+	)
+	if err != nil {
+		return fmt.Errorf("diff: %w", err)
+	}
+	d.changedDigestPaths[path] = fileNodeDiff{
+		path: fromNode.Path(),
+		from: fromNode.Digest(),
+		to:   toNode.Digest(),
+		diff: string(diffData),
+	}
+	return nil
+}
+
+func (d *manifestDiff) String() string {
+	var sb strings.Builder //nolint:varnamelen // sb is a common builder varname
+	if len(d.removedPaths) > 0 {
+		sb.WriteString("# Removed:\n\n")
+		sb.WriteString("```diff\n")
+		sortedPaths := slicesext.MapKeysToSortedSlice(d.removedPaths)
+		for _, path := range sortedPaths {
+			sb.WriteString("- " + d.removedPaths[path].String() + "\n")
+		}
+		sb.WriteString("```\n")
+		sb.WriteString("\n")
+	}
+	if len(d.addedPaths) > 0 {
+		sb.WriteString("# Added:\n\n")
+		sb.WriteString("```diff\n")
+		sortedPaths := slicesext.MapKeysToSortedSlice(d.addedPaths)
+		for _, path := range sortedPaths {
+			sb.WriteString("+ " + d.addedPaths[path].String() + "\n")
+		}
+		sb.WriteString("```\n")
+		sb.WriteString("\n")
+	}
+	if len(d.changedDigestPaths) > 0 {
+		sb.WriteString("# Changed content:\n\n")
+		sortedPaths := slicesext.MapKeysToSortedSlice(d.changedDigestPaths)
+		for _, path := range sortedPaths {
+			fnDiff := d.changedDigestPaths[path]
+			sb.WriteString("## `" + fnDiff.path + "`:\n")
+			sb.WriteString(
+				"```diff\n" +
+					fnDiff.diff + "\n" +
+					"```\n",
+			)
+		}
+		sb.WriteString("\n")
+	}
+	sb.WriteString(fmt.Sprintf(
+		"Total changes: %d (%d deletions, %d additions, %d changed)\n",
```

Comment:
This should perhaps just be for markdown format or a log statement.

Reply:
This could go in the unified diff format, at the very top, similar to a GH PR summary.

Reply:
Done